service/src/main/java/dk/dbc/search/solrdocstore/asyncjob/AsyncJobRunner.java

```diff
@@ -0,0 +1,138 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * dbc-solr-doc-store-service is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package dk.dbc.search.solrdocstore.asyncjob;
+
+import dk.dbc.search.solrdocstore.Config;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.stream.Collectors;
+import javax.annotation.PreDestroy;
+import javax.annotation.Resource;
+import javax.ejb.Lock;
+import javax.ejb.Singleton;
+import javax.ejb.Startup;
+import javax.enterprise.concurrent.ManagedExecutorService;
+import javax.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.time.Instant.now;
+import static javax.ejb.LockType.READ;
+
+/**
+ *
+ * @author DBC {@literal <dbc.dk>}
+ */
+@Singleton
+@Startup
+@Lock(READ)
+public class AsyncJobRunner {
+
+    @Inject
+    Config config;
+
+    @Resource
+    ManagedExecutorService mes;
+
+    private static final Logger log = LoggerFactory.getLogger(AsyncJobRunner.class);
+
+    private final ConcurrentSkipListMap<UUID, AsyncJobHandle> jobs;
+
+    public AsyncJobRunner() {
+        this.jobs = new ConcurrentSkipListMap<>();
+    }
+
+    /**
+     * Clean up jobs before undeploy
+     */
+    @PreDestroy
+    public void destroy() {
+        log.info("Shutting down jobs");
+        for (AsyncJobHandle handle : jobs.values()) {
+            if (handle.isStarted()) {
+                log.info("Canceling job: {}", handle.getJob().getName());
+                handle.cancel();
+
+            }
+            if (handle.isStarted()) {
+                log.info("Removing log for job: {}", handle.getJob().getName());
+                handle.getJob().removeLog();
+            }
+        }
+    }
+
+    public Map<String, String> jobs() {
+        return jobs.entrySet().stream()
+                .collect(Collectors.toMap(e -> e.getKey().toString(),
+                                          e -> e.getValue().getJob().getName()));
+    }
+
+    /**
+     * Get a job id and queue job for processing
+     *
+     * @param job async job
+     * @return handle
+     */
+    public String start(AsyncJob job) {
+        prune();
+        UUID id;
+        do {
+            id = UUID.randomUUID();
+        } while (jobs.containsKey(id));
+        AsyncJobHandle wrapper = new AsyncJobHandle(job);
+        jobs.put(id, wrapper);
```

Comment:
Lille detalje, men en race condition er vel egentlig mulig her. Mellem while loop og jobs.put(...) kan en anden thread vel egentlig nå at snuppe det samme id. Chancen er selvfølgelig lille, men det er det vel også for at randomUUID rammer et eksisterende job (og derfor laver du et while loop)? Måske blot synchronize metoden, den er vel ikke performance kritisk?

Reply:
ok